### PR TITLE
Remove second lifetime parameter from DataProvider

### LINKED ISSUE #752: Decide on the fate of the Borrowed variant in DataPayload
The Borrowed variant of DataPayload[Inner] pulls in the `'d` lifetime parameter.  This has caused some challenges, including:

- Less ergonomic (extra lifetime parameter)
- Issues related to trait objects (see comment in erased.rs)

The main use case for the Borrowed variant is to pave the way toward future no-alloc use cases.

If we decided to keep the Borrowed variant, there are some options we could explore to work around the trait object issues mentioned above.

Also decide on the names of the lifetimes (`'d` and `'s`).

### PR DESCRIPTION
**Pull request by sffc**
Fixes #752

### CODE REVIEW

**Review by zbraniecki — APPROVED** [review]
lgtm!

File: provider/core/README.md [hunk 1]
```diff
@@ -41,16 +41,9 @@ This crate also contains some concrete implementations for testing purposes:
...
-
-- `'d` is the lifetime of a reference to the whole data struct being borrowed.
-- `'s` is the lifetime of fields within a data struct.
-
-Data structs and [`DataMarker`] use the `'s` lifetime, and this is also the lifetime that the
-[`Yokeable`] implementation should handle.
-
-`'s` should exceed `'d` (i.e., `'s: 'd`). If a data struct is fully owned, `'s: 'static`.
+Most [`DataProvider`] traits take one lifetime argument: `'data`. This represents the lifetime
+of fields within a data struct. [`DataMarker`] also takes the `'data` lifetime, and this is
```
Comment (Manishearth) [inline provider/core/README.md:45]: I think we should expand this to mention the cart, `'data` is the lifetime of data in the _cart_, which the data struct can borrow from.

File: ffi/capi/src/decimal.rs [hunk 2]
```diff
@@ -14,13 +14,13 @@ use icu_decimal::FixedDecimalFormat;
 /// Opaque type for use behind a pointer, is [`FixedDecimalFormat`]
 ///
 /// Can be obtained via [`icu4x_fixed_decimal_format_create()`] and destroyed via [`icu4x_fixed_decimal_format_destroy()`]
-pub type ICU4XFixedDecimalFormat<'d> = FixedDecimalFormat<'d, 'static>;
+pub type ICU4XFixedDecimalFormat<'data> = FixedDecimalFormat<'data>;
```
Comment (Manishearth) [inline ffi/capi/src/decimal.rs:17]: I think this should lose the lifetime entirely, actually. I don't think we should expose non-`'static` data providers across FFI just yet.

**Review by Manishearth — APPROVED** [review]

File: ffi/capi/src/decimal.rs (hunk 2 above)
Comment (sffc) [reply ffi/capi/src/decimal.rs:17]: I'd like to check in the PR as-is, and figure this out in a follow-up PR.

File: provider/core/README.md (hunk 1 above)
Comment (sffc) [reply provider/core/README.md:45]: Will follow up with this and some more cleanups

**Pull request merged**